wit/producer.wit

```diff
@@ -1,5 +1,7 @@
+/// The producer interface is used to send messages to a channel/topic.
 interface producer {
-    use messaging-types.{client, channel, message, error};
+    use types.{client, message, error};
     
-    send: func(c: client, ch: channel, m: list<message>) -> result<_, error>;
+    /// Sends the message using the given client.
+    send: func(c: borrow<client>, m: message) -> result<_, error>;
```

Comment:
The `topic` should be an argument of the `send` function. I would also make the argument names more revealing:

```
func(client: borrow<client>, topic: topic, message: message)
```

Reply:
Addressed making the argument names more revealing across all functions ✅ —as for the topic name being part of the parameter, what if we then have a discrepancy between the message's topic and the provided topic? I feel like this adds ambiguity for implementations that could hurt the portability of the interface.

Reply:
I'd argue for moving the `topic` from the `message` to the `send` function from two different sides: the domain modeling and the implementation side.

On the domain model side, a message should be independent from the (possibly many) topics it is published on. I should be able to send the message "Fire!" to `floors.first`, `floors.second`, and `floors.third` without changing or cloning it.

On the implementation side, backend systems have different semantics:

Kafka most closely resembles the original intent but introduces a `ProducerRecord` that is very closely tied to the semantics of the producer (in contrast to a message):

```
ProducerRecord<String, String> record = new ProducerRecord<>("floors.first", "Fire!");
producer.send(record);
```

RabbitMQ (`routing_key` as topic):

```
channel.basic_publish(exchange='alarms',
                      routing_key='floors.first',
                      body=message)
```

NATS (Rust client; subject = topic)

```
let subject = "floors.first";
let data = Bytes::from("Fire!");
client.publish(subject, data.clone()).await?;
```

Reply:
> without changing or cloning it.

I think that is where the disconnection is happening, should they all be the `message`? Or should you have `n` message, one for each `send` if you know the topic is different?

Reply:
Addressed ✅ —Added `topic` to send, @jocrau